Sbp privatereg airgap

lproven opened the pull request:
As requested, I have imported the air-gapped Private Registry installation guide from:
https://confluence.suse.com/display/ENGCTNRSTORY/Using+k3s+as+air-gapped+environment+for+SUSE+Private+Registry

I converted it to AsciiDoc using Pandoc, have cleaned up and edited it to the best of my ability. This is my first significant piece of work using ADoc.

I have included the single resulting file as-is, and included it directly after the Installation section.

I am sure this needs further work but it's already taken 2 days longer than expected!

r0ckarong commented:
Sorry wrong button, didn't mean to close.

dirkmueller commented:
@jsuchome Any comments from your side?

asettle reviewed (approved):
LGTM as it's a copy :) thanks for doing this, @lproven 

Really need @jsuchome to go through this and ensure it's technically accurate.

jsuchome commented on adoc/Private-Registry-Airgap.adoc:
```diff
@@ -0,0 +1,247 @@
+sudo chown +x /usr/local/bin/k3s
+----
+
+.. Install k3s. Do not start the service yet.
```
This actually is not a requirement, I have used INSTALL_K3S_SKIP_ENABLE during my testing, but there's no reason users should avoid starting k3s as a systemd service...

lproven replied:
OK, so, what should it say?

jsuchome commented on adoc/Private-Registry-Airgap.adoc:
```diff
@@ -0,0 +1,247 @@
+INSTALL_K3S_SKIP_ENABLE=true INSTALL_K3S_SKIP_DOWNLOAD=true ./install.sh
+----
+
+.. Fetch the images required for the SUSE Private Registry and install them
```
As you can see, these images are downloaded from the internet. So it might be better to move the section where the images are downloaded more up, before we move to the airgapped machine...

jsuchome replied:
1. Basically so that the parts where one is fetching stuff from internet goes first and then goes the other part when working in airgap environment. In the original text this is kind of mixed (first k3s stuff download + install, then harbor images, download + copy).

2. Yes, that's a good idea, point to the latest release, or just release that user wants to use.

lproven replied:
I have added a mention of checking, and I've updated the URL to point to `v1.18.13+k3s1`. Since I started this, it looks like v1.19 has been released too, but I do not know if k3s uses a minor_ver=even → stable, minor_ver=odd → unstable system or anything like that.

I have not otherwise disentangled the installation steps, as I have not installed this myself and do not have a suitable machine. 

My understanding of air-gapped deployment means that the host machine has no internet connection, so I do not know how the step that does `docker pull` in a `do` loop will work. Perhaps I misunderstand what air-gapped means in this context. 

@jsuchome, if you want the steps disentangled, could you please sent a PR or at least give me new numbered steps? I know nothing about k3s, SPRpbH or anything to do with it. I had never heard of them before I was asked to migrate your Howto from Confluence.

jsuchome replied:
_My understanding of air-gapped deployment means that the host machine has no internet connection, so I do not know how the step that does docker pull in a do loop will work. Perhaps I misunderstand what air-gapped means in this context._

That was exactly my point - and now you can see that the original document is poorly written when it was not clear from it. So, the download parts need to be done on a machine with internet access, that's why I'm suggesting to move `docker pull` commands up to the top where you are also downloading `k3s` binary and installation script. 

We're saying _Transfer both files to your air-gapped machine._ about k3s above, so we should move the `docker pull` commands before this, and then explain something like _transfer k3s files and harbor docker images to your air-gapped machine_

jsuchome replied:
About the version: I would not care that much, linking specific version from this document - after all, the link becomes obsolete when they release a new one. Just say to fetch the latest stable version from the list of releases...

lproven replied:
OK, done. I have said use the latest stable version, linked to the releases page https://github.com/k3s-io/k3s/releases and the latest-release page https://github.com/k3s-io/k3s/releases/latest and removed the reference to any specific version. 

I have removed the "for example" mention when downloading the install script, and specified the filename for the `k3s` binary.

jsuchome commented on adoc/Private-Registry-Airgap.adoc:
```diff
@@ -0,0 +1,247 @@
+ image="registry.suse.com/harbor/harbor-$component:2.1.0"; \
+ docker pull $image; docker save -o "images/harbor-$component-2.1.0.tar" $image; \
+done
+sudo cp images/*.tar /var/lib/rancher/k3s/agent/images/
```
Here, better state explicitly to _move the image tar files into the airgap machine under /var/lib/rancher/k3s/agent/images/ directory_

lproven replied:
ISTM that the `copy` command in the later steps _requires_ that the tar files are somewhere else, as it copies them _to_ the specified directory. No?

jsuchome replied:
My point is that that with current list of commands, it suggests that you download the images and immediately, on the same system copy them into the right directory.

However we're talking about the case where k3s is actually running in the airgap environment - that means limited, or even no internet access. So I'm suggesting to split those commands so it is clear that
1. the images are downloaded on the machine with full internet access
2. later they are copied to the machine where k3s should run - and we do not need to suggest the way how end users will do it, they could indeed copy it on the same machine (which was for example temporary give internet access just for the download) or some physical device like USB stick might be needed to transport the files.

lproven replied:
I think I have addressed this now. Please could you give it a read-through and say yes or no?

lproven replied:
*bump*

jsuchome commented on adoc/Private-Registry-Airgap.adoc:
```diff
@@ -0,0 +1,247 @@
+[source,bash]
+----
+su
+k3s server
```
As mentioned above, `systemctl k3s start` should work (though I did not try it this way)

jsuchome commented:
I think it is technically correct ... as it is the way I wrote it :-) Just added some minor comments.

My issue with this is that it was not exactly written as a documentation bit, more like a developer how-to. So I'd think it could benefit from some style polishing.

lproven commented:
> I've added a commit fixing some things. Mainly:
> 
>     * This is adding "a guide inside a guide" so a section heading must be added and all following headings within the file moved down one level. It previously started on `Introduction` and this was also the less than useful point in the TOC. 

OK, fine.

> I'm not happy with the heading as it is but I also don't know a better way to spell out "air gap environment".

I have slightly revised it. BTW we (meaning the "Stylewich" committee) have recently changed the guidance in the Style Guide, and it now only recommendes title case for book/volume/collection titles -- **not** for section headings any more. This is not retroactive -- there's no need to go back and change all the old ones.

>     * Pandoc converts code example blocks to "format by indent" which (while valid) is a major headache to debug and fix in Asciidoc, I would suggest to always use the full code block syntax using `----` tags and not indent beyond how the example was originally indented.

OK. This is news to me; I only knew about formatting by indent. I probably removed some such markup.

>     * If possible, use the `[source,<LANG>]` formatting for all code blocks, this will give highlighting in your editor if it supports AsciiDoc and better auto-highlighting in the final document.

I was going to say that I tried that and it didn't work for me -- but your version doesn't work for me, either!

Firstly, I did a lot of Googling and it was not clear what "language" to specify for shell snippets: "shell" or "bash" or what. I tried both; in both cases, the result was a grey-tinted empty block. 

So I removed it again.

This is what happens with your formatting too, for me.

The YAML ones work fine. I do not know if it's because it's YAML or because they have filenames.

I am using Atom with the AsciiDoctor Preview plugin for ADoc stuff -- but as I said, I almost never use ADoc and this is my first ever ADoc commit to any of our docs.

> Unrelated: I've removed a wrong leveloffset for the Troubleshooting include.

OK

jsuchome commented:
I think my comments were not addressed.

jsuchome commented on adoc/Private-Registry-Airgap.adoc:
```diff
@@ -0,0 +1,254 @@
+sudo mkdir -p /var/lib/rancher/k3s/agent/images/
+sudo cp k3s-airgap-images-amd64.tar /var/lib/rancher/k3s/agent/images/
+sudo cp k3s /usr/local/bin/k3s
+sudo chown +x /usr/local/bin/k3s
```
And the install script:
`sudo chown +x install.sh`

jsuchome commented on adoc/Private-Registry-Airgap.adoc:
```diff
@@ -0,0 +1,254 @@
+some optional features (such as the Traefik ingress controller) here:
+https://github.com/rancher/k3s/releases/download/v1.18.13%2Bk3s1/k3s-airgap-images-amd64.tar[k3s-airgap-images-amd64.tar]
++
+Transfer both files to your air-gapped machine.
```
Yeah, it's actually 3 files: k3s binary, k3s-airgap-images and install.sh

jsuchome commented:
I updated the confluence page, now it should reflect the order of actions correctly.

lproven reviewed (commented):
> I updated the confluence page, now it should reflect the order of actions correctly.

Thanks. I have added a couple of bits of extra text from it.

jsuchome commented on adoc/Private-Registry-Airgap.adoc:
```diff
@@ -0,0 +1,271 @@
+.. Install k3s. Do not start the service yet. (The `INSTALL_K3S_SKIP_ENABLE`
+setting means that the `systemd` service for k3s will not be enabled and
+started, but you can ignore the value and actually use the service.)
++
```
Just make it more clear that it's optional, like 

_Install k3s. It is not necessary to start the systemd service yet, if you want to watch the start process closely for the first time_

lproven replied:
Done. I have moved the explanation of `INSTALL_K3S_SKIP_ENABLE` into a tip; is that OK?

jsuchome commented on adoc/Private-Registry-Airgap.adoc:
```diff
@@ -0,0 +1,271 @@
+mkdir images
+for component in core nginx notary-server notary-signer trivy-adapter registryctl jobservice registry db redis; do
+ image="registry.suse.com/harbor/harbor-$component:2.1.0"; \
+ docker pull $image; docker save -o "images/harbor-$component-2.1.0.tar" $image; \
```
Please replace `2.1.0` with `2.1.1`

dirkmueller replied:
or maybe do :2.1, which will automatically update?

lproven replied:
I have changed this to use 2.1.2 but I have no way to test this, so please verify that I have got it right.

lproven replied:
I also note that what looks to me like a version number is embedded here:

`export HELM_EXPERIMENTAL_OCI=1`
`helm chart pull registry.suse.com/harbor/harbor:1.5`
`helm chart export registry.suse.com/harbor/harbor:1.5`

Does that need to be 2.1.2 as well, @jsuchome ?

jsuchome replied:
Nope, 1.5 is correct here. Version of the helm chart is different than the versions of images
